### PR TITLE
ReactActivityDelegate implements LifecycleObserver

### PR DESCRIPTION
**Pull request by dulmandakh**
## Summary

Android introduced Lifecycles and LifecycleObserver to make it easy to respond to component lifecycle events. This PR changes ReactActivityDelegate to implement LifecycleObserver and handle lifecycle.

## Changelog

[Android] [Changed] - ReactActivityDelegate implements LifecycleObserver

## Test Plan

CI is green and RNTester is working just fine

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java [hunk 1]
```diff
@@ -94,12 +93,14 @@ protected void loadApp(String appKey) {
...
   }
 
+  @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
   protected void onPause() {
     if (getReactNativeHost().hasInstance()) {
       getReactNativeHost().getReactInstanceManager().onHostPause(getPlainActivity());
     }
   }
 
+  @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
```
Comment (benoitdion) [inline ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: `ON_RESUME`?

**Comment by cpojer** [comment]
@dulmandakh: What's the status of this PR?

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java [hunk 2]
```diff
@@ -13,37 +13,36 @@
...
   private @Nullable ReactRootView mReactRootView;
   private @Nullable DoubleTapReloadRecognizer mDoubleTapReloadRecognizer;
   private @Nullable PermissionListener mPermissionListener;
   private @Nullable Callback mPermissionsCallback;
 
-  @Deprecated
-  public ReactActivityDelegate(Activity activity, @Nullable String mainComponentName) {
-    mActivity = activity;
-    mMainComponentName = mainComponentName;
-  }
```
Comment (makovkastar) [inline ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: Is it safe to delete this constructor now? It caused a few compile errors in the FB codebase, but I fixed them. How about other people's code?

Comment (dulmandakh) [reply ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: it's been deprecated for a while, also activities must extend ReactActivity and everything will work fine.

**Comment by makovkastar** [comment]
After looking at this PR once again, I'm a bit concerned that now we have ReactActivityDelegate lifecycle events sent from two different places: we still call `mDelegate.onCreate(savedInstanceState)` from ReactActivity and we rely on the lifecycle events reported by the lifecycle owner. @dulmandakh what do you think?

**Comment by makovkastar** [comment]
@dulmandakh could you please check my previous comment? I would like to know what you think, because I'm leaning towards keeping the current implementation and not using `LifecycleObserver` for `ReactActivityDelegate`.

**Comment by dulmandakh** [comment]
@makovkastar ok. I'm closing this PR

**Pull request closed without merging**